### PR TITLE
APEX (Asynchronous Program EXecution) design

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

Fork choice doesn't need to depend on executing of all of the state transitions, and this dependency currently slows down the network when demand spikes.

#### Summary of Changes

Design proposal that isolates vote program from all the other programs.

Fixes #

tag @carllin @buffalu @jstarry @t-nelson @jackcmay @pgarg66 @sakridge

### CODE REVIEW

File: docs/src/proposals/ape.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
+== Asynchronous Program Execution ==
```
Comment (buffalu) [inline docs/src/proposals/ape.md:1]: would be helpful for me + others to elaborate on what problem this solves.

File: docs/src/proposals/ape.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
+== Asynchronous Program Execution == 
+
+Execution of state for fork choice can run asynchronously w.r.t program execution and computing the BankHash.
+
+
+=== Isolation domains ===
+
+Accounts include an additional field, the isolation domain. Currently only two domains exist.
```
Comment (buffalu) [inline docs/src/proposals/ape.md:8]: there are 4 listed below :)

File: docs/src/proposals/ape.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+Computing the BankHash evalutes all the non VoteProgram instructions. Transactions that include VoteProgram and any non VoteProgram accounts fail immediatly and are ignored.  The LS computation and finalizing the IsolationProgram::MoveDomains are done at the epoch boundary for the N+1 epoch.  The BankHash computation must not fall behind the ForkHash by more then 1 epoch or the cluster will halt.
+
+=== VoteProgram::Vote ===
+
+The Vote instruction contains ForkHashes only. BankHash computation is done outside of the root that consensus validators vote on. Consider mixing in the epoch snapshot hash into PoH once an epoch, and nodes should continue to gossip snapshot hashes continuously to ensure they are not experiencing an inconsistency failure.
+
+=== Invalid Fee Payers ===
+
+Transactions that fail due to fee payers unable to pay for the execution are ignored.  This means that leaders could stuff blocks with invalid transactions up to the known compute limit.
```
Comment (buffalu) [inline docs/src/proposals/ape.md:56]: "This means that leaders could stuff blocks with invalid transactions up to the known compute limit"

if invalid txs, what is compute limit then? does this become more about "block size" in the literal sense on size of the block (sum of all serialized versioned txs payloads plus metadata?)

File: docs/src/proposals/ape.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+A transaction cannot write to a domain while reading any other domain.  Any transaction that attempt to do so fails. The Vote program is isolated such that any transactions that writes to a VoteProgram account can only read from VoteProgram domain, otherwise the transaction fails.
+
+=== Replay ===
+
+Replay of banks is split up into 2 modes ForkHash and BankHash.
+
+=== ForkHash ===
+
+Computing the ForkHash only evalutes the VoteProgram instructions.  Transactions that include VoteProgram and any non VoteProgram accounts fail immediatly and are ignored.
```
Comment (0xEdgar) [inline docs/src/proposals/ape.md:44]:
```suggestion
Computing the ForkHash only evaluates the VoteProgram instructions.  Transactions that include VoteProgram and any non VoteProgram accounts fail immediately and are ignored.
```

File: docs/src/proposals/ape.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+=== BankHash ===
+
+Computing the BankHash evalutes all the non VoteProgram instructions. Transactions that include VoteProgram and any non VoteProgram accounts fail immediatly and are ignored.  The LS computation and finalizing the IsolationProgram::MoveDomains are done at the epoch boundary for the N+1 epoch.  The BankHash computation must not fall behind the ForkHash by more then 1 epoch or the cluster will halt.
```
Comment (0xEdgar) [inline docs/src/proposals/ape.md:48]:
```suggestion
Computing the BankHash evaluates all the non VoteProgram instructions. Transactions that include VoteProgram and any non VoteProgram accounts fail immediately and are ignored.  The LS computation and finalizing the IsolationProgram::MoveDomains are done at the epoch boundary for the N+1 epoch.  The BankHash computation must not fall behind the ForkHash by more than 1 epoch or the cluster will halt.
```

**Review by t-nelson — COMMENTED** [review]
can we wrap at ~80char to ease commenting?

**Comment by t-nelson** [comment]
seems like a really invasive way to get to "remove consensus messages from the transaction pipeline" ?

**Comment by aeyakovenko** [comment]
> seems like a really invasive way to get to "remove consensus messages from the transaction pipeline"?

I think this is pretty low overhead. The failing txs that mix domains can be done at the tx processor. And the isolation program accounts are the only ones that can move domains and the only ones that keep track of that state.

**Comment by buffalu** [comment]
what reasons would someone need to move domains? it seems like just transferring SOL to vote accounts?

**Comment by aeyakovenko** [comment]
@t-nelson I think like NONCEs, this might need to be part of System.

File: docs/src/proposals/ape.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+AccountMeta domain fields;
+
+* domain - The isolation domain.
+
+* activation epoch - The epoch when this account is active in the
+domain.
+
+The isolation domain. Currently only two domains exist:
+
+* 0 - The default domain where all things live
```
Comment (mschneider) [inline docs/src/proposals/ape.md:29]: Ring 0 is

**Comment by ryoqun** [comment]
(seems this is superseded by https://github.com/solana-foundation/solana-improvement-documents/pull/45)

**Pull request closed without merging**